supervisor/compressors.py

```diff
@@ -0,0 +1,59 @@
+"""
+Compressor implementations which used on logging file rotation.
+"""
+
+import bz2
+import gzip
+import os
+
+try:
+    import lzma
+except ImportError:
+    from backports import lzma
+
+class Compressor:
+    def compress(self, src, tgt):
+        """Abstract method for compression"""
+        pass
+
+
+class CopyCompressor(Compressor):
+    """Just copy files, do not any compressions"""
+    def __init__(self):
+        self.suffix = ""
+
+    def compress(self, src, tgt):
+        os.rename(src, tgt)
+
+
+class GzCompressor(Compressor):
+    def __init__(self, compressionLevel=9):
+        self.suffix = ".gz"
+        self.compressionLevel = compressionLevel
+
+    def compress(self, src, tgt):
+        with open(src, "rb") as f_src:
+            with gzip.open(tgt + self.suffix, "wb", compresslevel=self.compressionLevel) as f_tgt:
+                f_tgt.writelines(f_src)
```

Comment:
Similar to [#619](https://github.com/Supervisor/supervisor/pull/619/files#r30990694), this kind of blocking code can't be used inside `supervisord`. It may take several seconds to compress the log file before control is returned to the event loop. During this time, `supervisord` won't be able to do any of its normal processing, e.g. it won't respond to commands from `supervisorctl`. 

Blocking in this way will also cause problems for the programs running under `supervisord`. If a program writes to its `stdout` file descriptor when `supervisord` is blocked compressing the log, `supervisord` will not be able to read the data from the pipe.  The pipe buffer can then become full.  If that happens, the program may hang in the write or the write may fail.

Reply:
Maybe this code could be reworked to make the compression part be asynchronous and not block the code ?